File: Javascript/Heap/M-2336. Smallest Number in Infinite Set.ts
class SmallestInfiniteSet {
  heap: number[];
  nums: Set<number>;
  constructor() {
    this.heap = new Array(1000).fill(null).map((_, key) => key + 1);
    this.nums = new Set(this.heap);
  }

  popSmallest(): number | null {
    if (this.heap.length === 0) return null;
    if (this.heap.length === 1) {
      const num = this.heap.pop() as number;
      this.nums.delete(num);
      return num;
    }

    const root = this.heap[0];
    this.heap[0] = this.heap.pop() as number;
    this.heapifyDown(0);
    this.nums.delete(root);
    return root;
  }

  addBack(num: number): void {
    if (!this.nums.has(num)) {
      this.nums.add(num);
      this.heap.push(num);
      this.heapifyUp(this.heap.length - 1);
    }
  }

  private heapifyDown(idx: number): void {
    const [leftIdx, rightIdx] = [idx * 2 + 1, idx * 2 + 2];
    let smallestIdx = idx;
    if (
      leftIdx < this.heap.length &&
      this.heap[leftIdx] < this.heap[smallestIdx]
    ) {
      smallestIdx = leftIdx;
    }
    if (
      rightIdx < this.heap.length &&
      this.heap[rightIdx] < this.heap[smallestIdx]
    ) {
      smallestIdx = rightIdx;
    }
    if (idx !== smallestIdx) {
      [this.heap[smallestIdx], this.heap[idx]] = [
        this.heap[idx],
        this.heap[smallestIdx],
      ];
      this.heapifyDown(smallestIdx);
    }
  }

  private heapifyUp(idx: number): void {
    const parentIdx = (idx - 1) >> 1; // 1, 2 -> 0 and 3, 4 -> 1
    if (parentIdx >= 0 && this.heap[parentIdx] > this.heap[idx]) {
      [this.heap[parentIdx], this.heap[idx]] = [
        this.heap[idx],
        this.heap[parentIdx],
      ];
      this.heapifyUp(parentIdx);
    }
  }
}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * var obj = new SmallestInfiniteSet()
 * var param_1 = obj.popSmallest()
 * obj.addBack(num)
 */

var obj = new SmallestInfiniteSet();

File: Heap/M-2336. Smallest Number in Infinite Set/M-2336. Smallest Number in Infinite Set.py
class SmallestInfiniteSet:
    def __init__(self):
        self.min_heap = [i + 1 for i in range(1000)]
        self.nums = set(self.min_heap)

    def popSmallest(self) -> int:
        if len(self.min_heap) == 0:
            return None

        if len(self.min_heap) == 1:
            root = self.min_heap.pop()
            self.nums.remove(root)
            return root

        root = self.min_heap[0]
        self.nums.remove(root)
        self.min_heap[0] = self.min_heap.pop()
        self.__heapify_down(0)
        return root

    def addBack(self, num: int) -> None:
        if num not in self.nums:
            self.nums.add(num)
            self.min_heap.append(num)
            self.__heapify_up(len(self.min_heap) - 1)

    def __heapify_up(self, idx: int) -> None:
        parent_idx = (idx - 1) // 2
        if parent_idx >= 0 and self.min_heap[parent_idx] > self.min_heap[idx]:
            self.min_heap[parent_idx], self.min_heap[idx] = (
                self.min_heap[idx],
                self.min_heap[parent_idx],
            )
            self.__heapify_up(parent_idx)

    def __heapify_down(self, idx: int) -> None:
        left_idx, right_idx = idx * 2 + 1, idx * 2 + 2
        smallest_idx = idx
        if (
            left_idx < len(self.min_heap)
            and self.min_heap[left_idx] < self.min_heap[smallest_idx]
        ):
            smallest_idx = left_idx
        if (
            right_idx < len(self.min_heap)
            and self.min_heap[right_idx] < self.min_heap[smallest_idx]
        ):
            smallest_idx = right_idx

        if smallest_idx != idx:
            self.min_heap[smallest_idx], self.min_heap[idx] = (
                self.min_heap[idx],
                self.min_heap[smallest_idx],
            )
            self.__heapify_down(smallest_idx)


# Your SmallestInfiniteSet object will be instantiated and called as such:
obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)


""" You can simply use heapq lib """
# import heapq

# class SmallestInfiniteSet:

#     def __init__(self):
#         self.nums = [i+1 for i in range(1000)]
#         heapq.heapify(self.nums)

#     def popSmallest(self) -> int:
#         return heapq.heappop(self.nums)


#     def addBack(self, num: int) -> None:
#         if num in self.nums:
#             return
#         heapq.heappush(self.nums, num)
